File: ruby/signing.rb
require 'uri'
require 'net/http'
require 'securerandom'
require 'openssl'
require 'jwt'

# Replace with your own values
EMBED_PATH   = 'your path here'
CLIENT_ID    = 'your clientid here'
EMBED_SECRET = 'your secret here'

def urlencode(pairs)
  URI.encode_www_form(pairs)
end

def generate_signed_embed_url()
  params = {
    ":nonce": SecureRandom.uuid,
    ":email": "xyz@xyz.com",
    ":external_user_id": "xyz@xyz.com",
    ":client_id": CLIENT_ID,
    ":time": Time.now.to_i,
    ":session_length": 3600,
    ":mode": "userbacked",
    ":external_user_team": "EmbedTeam",
    # custom controls/parameters
    # "Store-Region": "West",
  }

  url_with_params = "#{EMBED_PATH}?#{urlencode(params)}"

  signature = OpenSSL::HMAC.hexdigest('sha256', EMBED_SECRET, url_with_params)

  url_with_signature = "#{url_with_params}&:signature=#{signature}"
  return url_with_signature
end

def generate_jwt_embed_url()
  payload = {
    "iss": CLIENT_ID,
    "sub": "xyz@xyz.com",
    "jti": SecureRandom.uuid,
    "ver": "1.1",
    "aud": "sigmacomputing",
    "iat": Time.now.to_i,
    "exp": Time.now.to_i + 3600,
    "account_type": "Pro",
    "teams": ["EmbedTeam"],
  }
  token = JWT.encode(payload, EMBED_SECRET, 'HS256')
  return "https://app.sigmacomputing.com/<your-org>/<your-workbook>?:embed=true&:jwt=#{token}"
end

def main()
  url_with_signature = generate_signed_embed_url()
  puts "======= Signed Embed URL ======="
  puts url_with_signature
  puts "==============================="

  url_with_jwt = generate_jwt_embed_url()
  puts "======= JWT Embed URL ======="
  puts url_with_jwt
  puts "==============================="
end

main()

File: node/signing.js
const crypto = require("crypto");
const { URLSearchParams } = require("url");
const jwt = require("jsonwebtoken");

// Replace with your own values
const EMBED_PATH = "your path here";
const CLIENT_ID = "your clientid here";
const EMBED_SECRET = "your secret here";

function urlencode(pairs) {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(pairs)) {
    params.append(key, value);
  }
  return params.toString();
}

function generateSignedEmbedUrl() {
  const params = {
    ":nonce": crypto.randomBytes(16).toString("hex"),
    ":email": "xyz@xyz.com",
    ":external_user_id": "xyz@xyz.com",
    ":client_id": CLIENT_ID,
    ":time": Math.floor(new Date().getTime() / 1000),
    ":session_length": 3600,
    ":mode": "view",
    ":external_user_team": "EmbedTeam",
    // custom controls/parameters
    // "Store-Region": "West",
  };

  const urlWithParams = `${EMBED_PATH}?${urlencode(params)}`;

  const signature = crypto
    .createHmac("sha256", Buffer.from(EMBED_SECRET, "utf8"))
    .update(Buffer.from(urlWithParams, "utf8"))
    .digest("hex");

  const urlWithSignature = `${urlWithParams}&:signature=${signature}`;
  return urlWithSignature;
}

function generateJwtEmbedUrl() {
  const payload = {
    iss: CLIENT_ID,
    sub: "xyz@xyz.com",
    aud: "sigmacomputing",
    jti: crypto.randomBytes(16).toString("hex"),
    ver: "1.1",
    iat: Math.floor(new Date().getTime() / 1000),
    exp: Math.floor(new Date().getTime() / 1000) + 3600,
    teams: ["EmbedTeam"],
    account_type: "Pro",
  };

  const token = jwt.sign(payload, EMBED_SECRET, {
    algorithm: "HS256",
    keyid: CLIENT_ID,
  });
  const urlWithToken = `https://app.sigmacomputing.com/<your org>/<your workbook>?:embed=true&:jwt=${token}`;
  return urlWithToken;
}

function main() {
  const jwtUrl = generateJwtEmbedUrl();
  console.log(
    `===========JWT Embed URL============
    ${jwtUrl}
    ==========================================`
  );

  const urlWithSignature = generateSignedEmbedUrl();
  console.log(
    `===========Signed Embed URL============
    ${urlWithSignature}
    ==========================================`
  );
}

main();
